### PR TITLE
Add AnimateDiff T2V & Extract common utilities & update FA for SD2.0

### PR DESCRIPTION
**Pull request by SamitHuang**
Inference tested on: GPU 3090,  Ascend 910A, and 910B. Results look fine
TODO: training

### CODE REVIEW

File: mindone/utils/load_params.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import os
+from typing import List, Optional, Union
+
+import mindspore as ms
+import mindspore.nn as nn
+from mindspore import Parameter
+from mindspore import log as logger
+
+# from mindspore._checkparam import Validator
+from mindspore.train.serialization import _load_dismatch_prefix_params, _update_param
```
Comment (zhtmike) [inline mindone/utils/load_params.py:11]: if possible, do not use the internal interface, i.e. `_xxx()` from external package, it will be frequently changed without notification.

File: mindone/schedulers/READMD.md
Comment (zhtmike) [inline mindone/schedulers/READMD.md:1]: typo: README.md

File: mindone/schedulers/ddim_schedule.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+class DDIMScheduler(nn.Cell):
+    """
+    `DDIMScheduler` extends the denoising procedure introduced in denoising diffusion probabilistic models (DDPMs) with non-Markovian guidance.
+
+    This model is based on nn.Cell for accelearting diffusion step computation.
+    """
+
+    def __init__(
+        self,
```
Comment (zhtmike) [inline mindone/schedulers/ddim_schedule.py:75]: better to add type hint for the refactored APIs, which can be used for API doc in future

File: mindone/schedulers/ddim_schedule.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        return self.timesteps
+
+    def _get_variance(self, alpha_prod_t, alpha_prod_t_prev):
+        beta_prod_t = 1 - alpha_prod_t
+        beta_prod_t_prev = 1 - alpha_prod_t_prev
+        variance = (beta_prod_t_prev / beta_prod_t) * (1 - alpha_prod_t / alpha_prod_t_prev)
+        return variance
+
+    # step function for denoising
+    def construct(self, model_output, timestep, sample, num_inference_steps, mask=None):
```
Comment (zhtmike) [inline mindone/schedulers/ddim_schedule.py:201]: better to add type hint, especially for `model_output`, and `sample`,  can not guess what type it accepts

File: examples/animatediff/tests/test_save_video.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import sys
+
+import numpy as np
+from PIL import Image
+
+__dir__ = os.path.dirname(os.path.abspath(__file__))
+sys.path.insert(0, os.path.abspath(os.path.join(__dir__, "..")))
+from animate_diff.utils.util import save_videos
+
+img_dir = os.path.join(__dir__, "../../videocomposer/datasets/webvid5")
```
Comment (zhtmike) [inline examples/animatediff/tests/test_save_video.py:12]: possible to move test data to a standardalone folder?

File: mindone/utils/load_params.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        param_dict = checkpoint
+    else:
+        raise TypeError(f"unknown checkpoint type: {checkpoint}")
+
+    if param_dict:
+        if ignore_net_params_not_loaded:
+            filter = param_dict.keys()
+        else:
+            filter = None
+        param_not_load, ckpt_not_load = load_param_into_net_with_filter(net, param_dict, filter=filter)
```
Comment (hadipash) [inline mindone/utils/load_params.py:126]:
```suggestion
        if ignore_net_params_not_loaded:
            filter = [param.name for _, param in net.parameters_and_names() if param.name not in parameter_dict]
            for param in filter:
                parameter_dict.pop(param)
        param_not_load, ckpt_not_load = ms.load_param_into_net(net, param_dict)
```

File: mindone/utils/misc.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from inspect import isfunction
+
+import mindspore.ops as ops
+
+
+def extract_into_tensor(a, t, x_shape):
+    b = t.shape[0]
+    out = ops.GatherD()(a, -1, t)
+    return out.reshape(b, *((1,) * (len(x_shape) - 1)))
```
Comment (hadipash) [inline mindone/utils/misc.py:9]: Will be good to have docs here, as it's not clear what this function does.

File: mindone/schedulers/ddim_schedule.py (hunk 3 above)
Comment (SamitHuang) [reply mindone/schedulers/ddim_schedule.py]: Updated. Thanks for the advice.

File: mindone/schedulers/ddim_schedule.py (hunk 2 above)
Comment (SamitHuang) [reply mindone/schedulers/ddim_schedule.py:77]: Updated. Thanks for the advice.

File: examples/animatediff/tests/test_save_video.py (hunk 4 above)
Comment (SamitHuang) [reply examples/animatediff/tests/test_save_video.py:12]: it's to avoid enlarging the repo resize. for test script, reusing data from neighbor should be fine.

File: mindone/schedulers/READMD.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# Developer Guide: How to add a new scheduler
+
+The scheduler class should inherit from mindspore.nn.Cell, and provide two key methods:
+    1. `add_noise`: adding noise to previous (latent) sample for diffusion forward
+    2. `construct`: predict the previous sample based on model output (typically, predicted latent noise) and input sample, i.e. diffusion backward
```
[the remark this answers is not in the document]
Comment (SamitHuang) [reply mindone/schedulers/READMD.md:5]: I was considering it too. But diffusion schedulers are evolving fast, not sure setting a base scheduler will restrict the flexibility. Let's do it later after investigating more advanced schedulers

File: mindone/utils/misc.py (hunk 6 above)
Comment (SamitHuang) [reply mindone/utils/misc.py:22]: Good advice. Clarified in docstring.

File: mindone/utils/load_params.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+from mindspore import Parameter
+from mindspore import log as logger
+
+# from mindspore._checkparam import Validator
+from mindspore.train.serialization import _load_dismatch_prefix_params, _update_param
+
+
+def load_param_into_net_with_filter(
+    net: nn.Cell, parameter_dict: dict, strict_load: bool = False, filter: Optional[List] = None
+):
```
[the remark this answers is not in the document]
Comment (SamitHuang) [reply mindone/utils/load_params.py:16]: Not the same. It's to filter the parameters in the network rather than checkpoints for loading. For example, loading unet 2d checkpoint to UNet3D cell, hundreds of warnings will emerge if use `ms. load_param_into_net ()` which makes it hard to spot the real bug

File: mindone/utils/load_params.py (hunk 5 above)
Comment (SamitHuang) [reply mindone/utils/load_params.py:126]: As mentioned above, it's for filtering network param to be loaded instead

File: mindone/utils/config.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+from omegaconf import DictConfig, ListConfig, OmegaConf
+
+
+def str2bool(b):
+    if b.lower() not in ["false", "true"]:
+        raise Exception("Invalid Bool Value")
+    if b.lower() in ["false"]:
+        return False
+    return True
```
[the remark this answers is not in the document]
Comment (SamitHuang) [reply mindone/utils/config.py:12]: For some args like enable_flash_attention, we want to enable it by default. Add `disable_X` for each such kind of arg can be redundant.

**Review by zhtmike — APPROVED** [review]

File: examples/animatediff/README.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+```
+python text_to_video.py --config configs/prompts/v2/1-ToonYou-MotionLoRA.yaml --L 16 --H 512 --W 512
+```
+
+By default, DDIM sampling is used, and the sampling speed is 1.07s/iter.
+
+Results using Zoom-In motion lora:
+
+<p float="left">
+<img src=https://github.com/SamitHuang/mindone/assets/8156835/f74280ca-a8c6-468a-909b-6189b6677c22 width="25%" />
```
Comment (wtomin) [inline examples/animatediff/README.md:79]: Similar problem with the GIF. It only plays once without replaying in infinite loops.

Comment (hadipash) [reply examples/animatediff/README.md:79]: Probably because it was saved to play only once.
https://github.com/mindspore-lab/mindone/pull/271/files#diff-343ec6ba8dc325aa1a32995c0b51e3524ad43c5564420b84e206df2bab1c721eR7

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+        dim_out = dim_out if dim_out is not None else dim
+
+        if activation_fn == "geglu":
+            act_fn = GEGLU(dim, inner_dim, approximate=False, dtype=dtype)
+        else:
+            raise NotImplementedError
+
+        self.net = nn.CellList([])
+        # project in
+        self.net.append(act_fn)
```
Comment (wtomin) [inline examples/animatediff/ad/modules/diffusionmodules/motion_module.py:109]: Maybe the input to this `act_fn` needs to be cast to fp32.

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+        # project dropout
+        self.net.append(nn.Dropout(p=dropout))
+        # project out
+        self.net.append(nn.Dense(inner_dim, dim_out).to_float(dtype))
+
+    def construct(self, hidden_states: ms.Tensor, scale: float = 1.0) -> ms.Tensor:
```
Comment (wtomin) [inline examples/animatediff/ad/modules/diffusionmodules/motion_module.py:115]: Is `scale` used here or somewhere else?

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+        self.dropout = ms.nn.Dropout(p=dropout)
+        position = np.expand_dims(np.arange(max_len), 1)
+
+        div_term = np.exp((np.arange(0, d_model, 2) * (-math.log(10000.0) / d_model)).astype(np.float32))
+        pe = np.zeros((1, max_len, d_model), dtype=np.float32)
+
+        pe[0, :, 0::2] = np.sin(position * div_term)
+        pe[0, :, 1::2] = np.cos(position * div_term)
+
+        self.pe = ms.Parameter(ms.Tensor(pe, dtype=ms.float32), name="pe", requires_grad=False)
```
Comment (wtomin) [inline examples/animatediff/ad/modules/diffusionmodules/motion_module.py:320]: Maybe using `ms.Tensor` instead of `ms.Parameter` is OK here. I have discussed this with Rustam before. It seems that current ms version does not require the nn.Cell's tensor attribute  to be a `ms.Parameter`.

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py [hunk 14]
```diff
@@ -0,0 +1,523 @@
...
+        if self.upcast_attention:
+            query = query.to_float(ms.float32)
+            key = key.to_float(ms.float32)
+
+        attention_scores = ms.ops.matmul(query, ms.ops.transpose(key, (0, 2, 1))) * self.scale
+
+        if attention_mask is not None:
+            attention_scores = attention_scores + attention_mask
+
+        if self.upcast_softmax:
```
Comment (wtomin) [inline examples/animatediff/ad/modules/diffusionmodules/motion_module.py:507]: Why not casting `attention_scores` to fp32 before adding it with `attention_mask`

File: examples/stable_diffusion_v2/inference/sd_infer.py [hunk 15]
```diff
@@ -42,7 +42,6 @@ def main(args):
         config,
         ckpt=config.model.pretrained_ckpt,
         use_lora=args.use_lora,
-        lora_rank=args.lora_rank,
```
Comment (wtomin) [inline examples/stable_diffusion_v2/inference/sd_infer.py:45]: Why deleting `lora_rank` here?

File: mindone/schedulers/ddim_schedule.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+    """
+    Rescales betas to have zero terminal SNR Based on https://arxiv.org/pdf/2305.08891.pdf (Algorithm 1)
+
+
+    Args:
+        betas (`torch.FloatTensor`):
+            the betas that the scheduler is being initialized with.
+
+    Returns:
+        `torch.FloatTensor`: rescaled betas with zero terminal SNR
```
Comment (wtomin) [inline mindone/schedulers/ddim_schedule.py:42]: `torch.FloatTensor` here maybe confusing. I suggest changing it to `ms.Tensor`

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py (hunk 13 above)
Comment (hadipash) [reply examples/animatediff/ad/modules/diffusionmodules/motion_module.py:320]: https://github.com/mindspore-lab/mindone/pull/248#discussion_r1427481430

File: mindone/visualize/videos.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
+import os
+
+import imageio
+import numpy as np
+
+
+def save_videos(frames: np.ndarray, path: str, fps=8, loop=1, concat=False):
```
Comment (hadipash) [inline mindone/visualize/videos.py:7]: Set `loop` to `0` by default to play GIFs indefinitely?

File: mindone/utils/params.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+                logger.info("Net params not loaded: {}".format([p for p in param_not_load if not p.startswith("adam")]))
+        logger.info("Checkpoint params not loaded: {}".format([p for p in ckpt_not_load if not p.startswith("adam")]))
+
+
+def count_params(model, verbose=False):
+    total_params = sum([param.size for param in model.get_parameters()])
+    trainable_params = sum([param.size for param in model.get_parameters() if param.requires_grad])
+
+    if verbose:
+        print(f"{model.__class__.__name__} has {total_params * 1.e-6:.2f} M params.")
```
Comment (hadipash) [inline mindone/utils/params.py:144]:
```suggestion
        logger.info(f"{model.__class__.__name__} has {total_params * 1.e-6:.2f} M params.")
```

File: examples/stable_diffusion_v2/inference/sd_infer.py (hunk 15 above)
Comment (SamitHuang) [reply examples/stable_diffusion_v2/inference/sd_infer.py:45]: No need. we infer rank_rank from checkpoint.

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py (hunk 13 above)
Comment (SamitHuang) [reply examples/animatediff/ad/modules/diffusionmodules/motion_module.py:320]: It's a little risky. Not sure this change is ok for different MS versions on Ascend/GPU

File: mindone/visualize/videos.py [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+        concat: if True and b>1, all videos will be concatnated in grids and saved as one gif.
+        loop: number of loops to play. If 0, it will play endlessly.
+    """
+    # input frames: (b f H W 3), normalized to [0, 1]
+    frames = (frames * 255).round().clip(0, 255).astype(np.uint8)
+    os.makedirs(os.path.dirname(path), exist_ok=True)
+
+    duration = 1 / fps
+    if len(frames.shape) == 4:
+        imageio.mimsave(path, frames, duration=duration)
```
Comment (hadipash) [inline mindone/visualize/videos.py:23]: imageio's core v2 is obsolete and [will be deleted in the future](https://imageio.readthedocs.io/en/stable/reference/userapi.html#writing-images).
Use this instead:
```python
import imageio.v3 as iio
iio.imwrite(...)    # passed parameters are the same
```

File: mindone/schedulers/ddim_schedule.py (hunk 16 above)
Comment (SamitHuang) [reply mindone/schedulers/ddim_schedule.py]: Fixed

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py (hunk 14 above)
Comment (SamitHuang) [reply examples/animatediff/ad/modules/diffusionmodules/motion_module.py:507]: should be fine. the precision is tested, mre < 10e-5

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py (hunk 11 above)
Comment (SamitHuang) [reply examples/animatediff/ad/modules/diffusionmodules/motion_module.py:109]: geglu should be fine. the tested mre error < 1e-4

File: examples/animatediff/README.md (hunk 10 above)
Comment (SamitHuang) [reply examples/animatediff/README.md]: updated

File: mindone/visualize/videos.py (hunk 19 above)
Comment (SamitHuang) [reply mindone/visualize/videos.py:23]: updated requirement

File: examples/animatediff/ad/modules/diffusionmodules/motion_module.py (hunk 12 above)
Comment (SamitHuang) [reply examples/animatediff/ad/modules/diffusionmodules/motion_module.py]: updated

**Review by wtomin — APPROVED** [review]
